### ISSUE TITLE
Get TaskCanceledException when calling Azure APIs

### ISSUE THREAD

**Issue by mjsabin01** [Azure/azure-sdk-for-net#4278]
I have noticed that in my App we have started to receive a large number of TaskCanceledException when using the managed APIs to call Azure endpoints.  From my logging I see about a 90 secs goes by and I get the exception.  Here is the stack trace:

`
System.Threading.Tasks.TaskCanceledException: A task was canceled.
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Rest.RetryDelegatingHandler.<>c__DisplayClass11_0.<<SendAsync>b__1>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Rest.RetryDelegatingHandler.<SendAsync>d__11.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Net.Http.HttpClient.<FinishSendAsync>d__58.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Azure.Management.ResourceManager.SubscriptionsOperations.<GetWithHttpMessagesAsync>d__6.MoveNext()
`

**Comment by vithal-wadje** [comment]
Same exception I am getting while getting the list of linked IoT hubs to the DPS  using package  Microsoft.Azure.Management.DeviceProvisioningServices in IotDpsClient get method.

**Comment by deathly809** [comment]
@mjsabin01 @vithal-wadje  Have you watched the network to see if the error is coming from Azure vs. an SDK issue?  Or are you saying that the SDK should output a better error?

**Comment by mjsabin01** [comment]
My errors are coming from a deployed Service Fabric cluster in Azure.    I only see it about a 10 times a day in my service. I guess the question what is the underlying root cause.  Does azure perform some throttling that would cause this to happen?

**Comment by deathly809** [comment]
@mjsabin01 The throttling error message you get is very specific, you would know.  Sounds like some type of exception on the service side of things.

**Comment by vithal-wadje** [comment]
@deathly809 Its coming from SDK itself , I have raised the issue in detail , if you get time please check & let me know if you have any solution on it [https://github.com/Azure/azure-sdk-for-net/issues/4287](url)

**Comment by deathly809** [comment]
I looked at the link.  Your error is coming from the RP, I am still not sure why you think this is on the SDK side.  Unless you are saying that the SDK has been created incorrectly?

**Comment by shahabhijeet** [comment]
@mjsabin01 can you check and let me know #4287 helps in resolving your issue.
If not please provide repro code for us to investigate this further.

**Comment by DanCaveman** [comment]
I am having this issue as well.  The following code I am running in a unit test and it starts out good for a hundred messages or so, then, it slows significantly as the threading errors roll in.  Any ideas what I am doing wrong?

`            var drainReceiver = new MessageReceiver(ProdIntegrationTestCnString, ProdTestQueue);
            drainReceiver.PrefetchCount = 100;
            var messageHandlerOptions = new MessageHandlerOptions((x) => Task.CompletedTask)
            {
                MaxConcurrentCalls = 1000,
                AutoComplete = false
            };

            _logger.Info("Starting to drain queue");

            drainReceiver.RegisterMessageHandler((message, cancellationToken) => drainReceiver.CompleteAsync(message.SystemProperties.LockToken), messageHandlerOptions);
            
            
            Thread.Sleep(Timeout.Infinite);    
        }`

**Comment by JamesFenton** [comment]
Out team upgraded from 1.7.0 to 1.21.0 of Microsoft.Azure.Management.Fluent and we started hitting repeated TaskCanceledException. The requests were not being rate limited. I managed to reproduce it in a simple console app that did this:
```
var azure = Azure
  .Configure()
  .Authenticate(credentials)
  .WithDefaultSubscription();
while (true)
{
    var scaleSets = azure.VirtualMachineScaleSets.ListByResourceGroup("MyResourceGroup");
    Thread.Sleep(5000);
}
```
It would work about 80% of the time and every now and again would throw TaskCanceledException. I noticed that it would throw the exception after about 100 seconds, which is the default timeout of `HttpClient`. 
Reverting back to version 1.11.1 fixed the problem for us.
